module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php

```diff
@@ -0,0 +1,107 @@
+use Ergonode\Product\Infrastructure\Grid\ProductGridBuilder;
+
+class FilteredQueryBuilderProvider implements FilteredQueryBuilderProviderInterface
+{
+    private DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory;
+
+    private ProductGridBuilder $gridBuilder;
+
+    private ProductIdsProvider $productIdsProvider;
+
+    private Connection $connection;
+
+    public function __construct(
+        DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory,
+        ProductGridBuilder $gridBuilder,
+        ProductIdsProvider $productIdsProvider,
+        Connection $connection
+    ) {
+        $this->productQueryBuilderFactory = $productQueryBuilderFactory;
+        $this->gridBuilder = $gridBuilder;
+        $this->productIdsProvider = $productIdsProvider;
+        $this->connection = $connection;
+    }
+
+
+    public function provide(?BatchActionFilter $filter): QueryBuilder
+    {
+        $queryBuilder = $this->getQueryBuilder();
+        if ($filter) {
+            $filterQuery = $filter->getQuery();
+            if ($filterQuery) {
+                if ($filter->getIds()) {
+                    if ($filter->getIds()->isIncluded()) {
+                        //query and include id
+                        $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                        return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->in('id', ':productsIds'))
+                            ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+                    }
+
+                    //query and exclude id
+                    $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                    return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->notIn('id', ':productsIds'))
+                        ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+                }
+
+                //only query
+                return $this->getFilteredQueryBuilder($filter->getQuery());
+            }
+
+            if ($filter->getIds()) {
+                if ($filter->getIds()->isIncluded()) {
+                    //only include ids
+                    return $queryBuilder->andWhere($queryBuilder->expr()->in('id', ':productsIds'))
+                        ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+                }
+
+                //only exclude id in all system
+                return $queryBuilder->andWhere($queryBuilder->expr()->notIn('id', ':productsIds'))
+                    ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+            }
+        }
+
+        //no filter or empty, get all items
+        return $queryBuilder;
+    }
+
+    private function getFilteredQueryBuilder(string $filter): QueryBuilder
+    {
+        $language = new Language('en_GB');
```

Comment:
```suggestion
        $language = $this->languageQuery->getRootLanguage();
```
Where 
`$languageQuery` instanceof `Ergonode\Core\Domain\Query\LanguageQueryInterface;`